### ISSUE TITLE
AssertionError in deeply nested class scenario (incremental with warm cache)

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#2931]
This is a small remainder of #2535 

Example to reproduce: run ``mypy -i main.py`` twice (without touching any files in between) where:
* main.py
```python
import ntcrash
```
* ntcrash.py
```python
from typing import NamedTuple

class C:
    def f(self) -> None:
        class D:
            class E:
                def g(self) -> None:
                    A = NamedTuple('A', [('x', int), ('y', int)])
                    self.a = A(1, 1)
        self.c = D.E().a
```
Possible fix would be to store all "anonymous" classes in ``globals``.

**Comment by ilevkivskyi** [comment]
One needs to add ``reveal_type(ntcrash.C().c)`` between two runs to trigger the crash.

**Comment by hauntsaninja** [comment]
I think this is fixed on master, possibly by #10431
